src/workerd/io/actor-sqlite.c++

```diff
@@ -662,6 +662,50 @@ kj::Maybe<kj::Promise<void>> ActorSqlite::onNoPendingFlush() {
   return outputGate.wait();
 }
 
+kj::Promise<kj::String> ActorSqlite::getCurrentBookmark() {
+  // This is an ersatz implementation that's good enough for local dev with D1's Session API.
+  //
+  // The returned bookmark satisfies the properties that D1 cares about:
+  //
+  // * Later bookmarks sort after earlier bookmarks.  We implement this by incrementing the bookmark
+  // * whenever getCurrentBookmark() is called.
+  //
+  // * Bookmarks from the current workerd session sort after bookmarks from previous sessions.  We
+  //   implement this by saving an ersatz bookmark in the SqliteMetadata table.
+
+  requireNotBroken();
+  uint64_t bookmark = 0;
+  KJ_IF_SOME(b, metadata.getLocalDevelopmentBookmark()) {
+    bookmark = b + 1;
+  }
+  metadata.setLocalDevelopmentBookmark(bookmark);
+
+  // TODO(cleanup): Left-padded number stringification should maybe be in KJ?
+  auto paddedHex = [](uint32_t n) {
+    kj::FixedArray<char, 8> result;
+    for (auto i = 0; i < result.size(); i++) {
+      char digit = n % 16;
+      n /= 16;
+      digit += digit < 10 ? '0' : ('a' - 10);
+      result[result.size() - 1 - i] = digit;
+    }
+    return result;
+  };
+
+  // Turn the bookmark into a format matching what Cloudflare's production returns.
+  constexpr uint32_t uint32_max = kj::maxValue;
+  kj::FixedArray<char, 32> pad;
+  pad.fill('0');
+  return kj::str(paddedHex(bookmark / uint32_max), '-', paddedHex(bookmark % uint32_max), '-',
```

Comment:
I guess with this algorithm people will probably never see a non-zero value for the first component.

Arguably we could instead increment the first component of the bookmark every time we reload the DO, and increment the second component for each call to `getBookmark()` within one session?

I'll leave it up to you though whether you think it's worth the effort.

Reply:
I'm going to merge it as is since we can always change it if we think it matters.